File: api/fmp.js
export default async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin", "*");

  const key = globalThis.process?.env?.FMP_KEY;
  if (!key) return res.json(null);

  // FMP uses bare ticker (no exchange prefix); strip any suffix like .TO, .V, .L
  const raw = (req.query.symbol ?? "").trim().toUpperCase();
  const symbol = raw.replace(/\.(TO|V|L|AX|NS|DE)$/i, "");
  if (!symbol) return res.status(400).json({ error: "Missing symbol" });

  try {
    const [metricsRes, incomeRes] = await Promise.all([
      fetch(`https://financialmodelingprep.com/api/v3/key-metrics/${encodeURIComponent(symbol)}?limit=1&apikey=${key}`),
      fetch(`https://financialmodelingprep.com/api/v3/income-statement/${encodeURIComponent(symbol)}?limit=4&apikey=${key}`),
    ]);

    const metrics = metricsRes.ok ? ((await metricsRes.json())?.[0] ?? null) : null;
    const income  = incomeRes.ok  ? ((await incomeRes.json())  ?? [])         : [];
    const latest  = income[0] ?? null;

    if (!metrics && !latest) return res.json(null);

    const n = (v, d = 2) => v != null && isFinite(v) ? +Number(v).toFixed(d) : null;

    return res.json({
      peRatio:          n(metrics?.peRatio, 1),
      evToEbitda:       n(metrics?.enterpriseValueMultiple, 1),
      psRatio:          n(metrics?.priceToSalesRatio, 2),
      roic:             metrics?.roic != null ? n(metrics.roic * 100, 1) : null,
      revenueAnnual:    latest?.revenue   != null ? n(latest.revenue   / 1e9, 2) : null,
      netIncomeAnnual:  latest?.netIncome != null ? n(latest.netIncome / 1e9, 2) : null,
      grossMargin:      latest?.grossProfitRatio != null ? n(latest.grossProfitRatio * 100, 1) : null,
      epsAnnual:        n(latest?.eps, 2),
      revenueGrowthYoY: income.length >= 2 && income[0]?.revenue && income[1]?.revenue
        ? n((income[0].revenue - income[1].revenue) / Math.abs(income[1].revenue) * 100, 1)
        : null,
      earningsHistory: income.slice(0, 4)
        .map(q => ({ period: q.period ?? q.date?.slice(0, 7) ?? "", eps: q.eps ?? null, revenue: q.revenue ?? null }))
        .filter(q => q.eps != null && q.revenue != null),
    });
  } catch (e) {
    return res.status(500).json({ error: e instanceof Error ? e.message : "FMP error" });
  }
}

File: backend/fetchers/fmp.py
"""Financial Modeling Prep fetcher."""

import httpx

from backend.config import get_settings

SOURCE = "fmp"
_BASE = "https://financialmodelingprep.com/api/v3"


class FMPFetcher:
    def __init__(self, client: httpx.AsyncClient) -> None:
        self._client = client

    @property
    def _key(self) -> str:
        return get_settings().fmp_key

    async def enrich(self, symbol: str) -> dict | None:
        if not self._key:
            return None
        bare = symbol.split(".")[0]

        def n(v, d=2):
            try:
                x = float(v)
                return round(x, d) if x == x else None
            except (TypeError, ValueError):
                return None

        try:
            import asyncio
            metrics_r, income_r = await asyncio.gather(
                self._client.get(f"{_BASE}/key-metrics/{bare}",
                                 params={"limit": 1, "apikey": self._key}),
                self._client.get(f"{_BASE}/income-statement/{bare}",
                                 params={"limit": 4, "apikey": self._key}),
            )
            metrics_list = metrics_r.json() if metrics_r.is_success else []
            income_list = income_r.json() if income_r.is_success else []
            metrics = metrics_list[0] if metrics_list else None
            latest = income_list[0] if income_list else None
            if not metrics and not latest:
                return None

            rev_growth = None
            if len(income_list) >= 2 and income_list[0].get("revenue") and income_list[1].get("revenue"):
                delta = income_list[0]["revenue"] - income_list[1]["revenue"]
                rev_growth = n(delta / abs(income_list[1]["revenue"]) * 100, 1)

            return {
                "peRatio": n(metrics.get("peRatio"), 1) if metrics else None,
                "evToEbitda": n(metrics.get("enterpriseValueMultiple"), 1) if metrics else None,
                "psRatio": n(metrics.get("priceToSalesRatio"), 2) if metrics else None,
                "roic": n(metrics.get("roic", 0) * 100, 1) if metrics and metrics.get("roic") is not None else None,
                "revenueAnnual": n(latest["revenue"] / 1e9, 2) if latest and latest.get("revenue") else None,
                "netIncomeAnnual": n(latest["netIncome"] / 1e9, 2) if latest and latest.get("netIncome") else None,
                "grossMargin": n(latest["grossProfitRatio"] * 100, 1) if latest and latest.get("grossProfitRatio") is not None else None,
                "epsAnnual": n(latest.get("eps"), 2) if latest else None,
                "revenueGrowthYoY": rev_growth,
                "earningsHistory": [
                    {"period": q.get("period") or (q.get("date") or "")[:7],
                     "eps": q.get("eps"), "revenue": q.get("revenue")}
                    for q in income_list[:4]
                    if q.get("eps") is not None and q.get("revenue") is not None
                ],
            }
        except Exception:
            return None
